### PR TITLE
Linux production build

### PR DESCRIPTION
**Pull request by ConorIA**
Provides a more polished experience on Linux. (#26)

Devs, please check the gruntfile. Using the 'release-prod' task makes a couple of references to 'Win', but I see that the same task is used for OSX. 

Currently I am using the official 256 icon, but it doesn't look too pretty. 

screenshot from 2016-09-02 17-31-30

### CODE REVIEW

File: electron/main.js [hunk 1]
```diff
@@ -221,7 +224,7 @@ function showMainWindow() {
   });
 
   main.webContents.on('crashed', function() {
-    main.reload();
+    process.crash();
```
Comment (lipis) [inline electron/main.js]: this should not be part of this PR

Comment (ConorIA) [inline electron/main.js]: Oops. You're right. This is the product of a sloppy merge from upstream on my part. I'll revert it.

**Comment by david-wire** [comment]
CLA approved

File: Gruntfile.coffee [hunk 2]
```diff
@@ -259,3 +259,4 @@ module.exports = (grunt) ->
   grunt.registerTask 'win-prod',  ['clean:win', 'update-keys', 'release-prod', 'electron:win_prod', 'create-windows-installer:prod']
 
   grunt.registerTask 'linux',     ['clean:linux', 'release-internal', 'electron:linux']
```
Comment (lipis) [inline Gruntfile.coffee]: `update-keys` here as well :)

File: Gruntfile.coffee [hunk 3]
```diff
@@ -259,3 +259,4 @@ module.exports = (grunt) ->
...
+  grunt.registerTask 'lin-prod',  ['clean:linux', 'update-keys', 'release-prod', 'electron:linux']
```
Comment (lipis) [inline Gruntfile.coffee]: `linux-prod` and align them all leaving one space to the longest name :)

**Comment by lipis** [comment]
Looks good here.. would it be better with 128px?
screen shot 2016-09-07 at 09 59 48

**Comment by ConorIA** [comment]
The icon still has a sharp black line on my Gnome installation, even at 128 px. 

screenshot from 2016-09-07 11-36-58

It is not as noticable at 64px. 

screenshot from 2016-09-07 11-44-09

That being said, I've looked at other icons on my system and most are >=256px, so the issue is probably more with Gnome than it is with the Wire icon.

**Comment by lipis** [comment]
We might go ahead with this version and tweak in the future.. Thank you for testing it..

**Comment by ConorIA** [comment]
Great! Cheers. Once you folks merge this and tag a release, I'll more the PKGBUILD I maintain back to this repo. 😄

File: electron/main.js [hunk 4]
```diff
@@ -128,6 +128,9 @@ ipcMain.on('google-auth-request', function(event) {
 ///////////////////////////////////////////////////////////////////////////////
 // APP Windows
 ///////////////////////////////////////////////////////////////////////////////
+
+const iconExt = (process.platform === 'win32') ? 'ico' : 'png';
```
Comment (lipis) [inline electron/main.js]: on a second thought we could make it better :) add that on line 44 and use the `ICON_PATH` instead..

```js
const ICON = 'wire.' + ((process.platform === 'win32') ? 'ico' : 'png');
const ICON_PATH = path.join(APP_PATH, 'img', ICON);
```

**Comment by ConorIA** [comment]
I think that should do it. 

Would it be a bad idea to condense this to one constant as: 
```
const ICON = path.join(APP_PATH, 'img', ('wire.' + ((process.platform === 'win32') ? 'ico' : 'png'));
```

**Comment by lipis** [comment]
The line is getting too long.. :/

File: electron/main.js [hunk 5]
```diff
@@ -128,6 +130,7 @@ ipcMain.on('google-auth-request', function(event) {
 ///////////////////////////////////////////////////////////////////////////////
 // APP Windows
 ///////////////////////////////////////////////////////////////////////////////
+
```
Comment (lipis) [inline electron/main.js]: let's delete that and I'll merge it..

**Comment by ConorIA** [comment]
Good to go!

**Comment by lipis** [comment]
🎆

**Comment by ryanleesipes** [comment]
When do we see this pop up on the site?

**Comment by lipis** [comment]
Our biggest issue is the [automatic updates](http://stackoverflow.com/q/38129722/8418) and how to tackle them on different distributions.. what's your take on that?

**Comment by ConorIA** [comment]
Perhaps you could just put a link on the website that would lead to a wiki page here on Github. That wiki page could include a list of community-maintained Linux packages, along with a reminder that Linux isn't officially supported (though I wish you guys would reconsider officialy Linux support!) For Arch-based distros, that would be my PKGBUILD: https://aur.archlinux.org/packages/wire-desktop/. 

I am sure someone out there is already making a package for Ubuntu, or it might be as easy as asking Travis to follow the instructions over here: http://www.titus-stahl.de/blog/2016/08/04/creating-an-ubuntu-package-of-wire-an-encrypted-messenger/. 

Heck, I wonder how easy it would be to get Travis to automatically update the PKGBUILD and thus automatically update the Arch package.

**Comment by ConorIA** [comment]
I'm not really willing to figure out how to set-up and update a PPA, but I threw together a quick script that uses docker to produce Ubuntu packages for Wire Desktop without any real user interaction. Not sure if it would be of any use to someone who would like to maintain an Ubuntu PPA. 

https://gist.github.com/ConorIA/21a192912e7f1ec40fde26a854e4ad2c

**Comment by lipis** [comment]
Thank you @ConorIA for spending the time and researching the topic! We are definitely interested..

**Comment by ConorIA** [comment]
The first version of the Gist I uploaded has a couple build deps missing. It's improved now. Here is the output. If someone with Ubuntu can test these images, we're off to an alright start. 

http://cubert.conr.ca:8001/d/45b7bf0cd3/

**Comment by ConorIA** [comment]
I did a little but more poking around to test whether we could make something like Travis do the legwork for us. I am a lot more familiar with the Girlab multirunner, so I mirrored my fork over on Gitlab and added a gitlab-ci.yml to see if I couldn't host a debian repository. To my surprise, it worked! So, while this is a little disjointed at the moment, what one could potentially configure, with a little more knowledge and patience is a deb repo that is rebuilt and updated every time you push a change to Github!

Here are the ingedients for my proof of concept: 
- The project is mirrored at https://gitlab.com/ConorIA/wire-tests
- When this project is updated, it will use the Gitlab multirunner ([config is here](https://gitlab.com/ConorIA/wire-tests/blob/master/.gitlab-ci.yml)) to build using [these commands](https://gitlab.com/snippets/26204).

The debian packages are published to the following repo: 
```
deb http://conoria.gitlab.io/wire-tests build/
```
For now there are a couple of drawbacks: 
- The main issue is that these build are not signed. In order to sign them, I guess I'd need to manually sign them, but that defeats the purpose of the automation. 
- I could only figure out how to add one package at a time, but it might be as easy as adding two repos, e.g. instead of "binary" use "i686" and "x86_64".

I got the working idea for this from [here](https://hiddentao.com/archives/2011/06/27/setting-up-a-simple-ubuntu-apt-repository/). I think that the [apt-now](https://github.com/cmotc/apt-now) tool might also be useful.

**Pull request merged**